pkg/manager/member/dm_master_upgrader.go

```diff
@@ -0,0 +1,150 @@
+		// Manually bypass tidb-operator to modify statefulset directly, such as modify dm-master statefulset's RollingUpdate straregy to OnDelete strategy,
+		// or set RollingUpdate to nil, skip tidb-operator's rolling update logic in order to speed up the upgrade in the test environment occasionally.
+		// If we encounter this situation, we will let the native statefulset controller do the upgrade completely, which may be unsafe for upgrading dm-master.
+		// Therefore, in the production environment, we should try to avoid modifying the dm-master statefulset update strategy directly.
+		newSet.Spec.UpdateStrategy = oldSet.Spec.UpdateStrategy
+		klog.Warningf("tidbcluster: [%s/%s] dm-master statefulset %s UpdateStrategy has been modified manually", ns, dcName, oldSet.GetName())
+		return nil
+	}
+
+	setUpgradePartition(newSet, *oldSet.Spec.UpdateStrategy.RollingUpdate.Partition)
+	podOrdinals := helper.GetPodOrdinals(*oldSet.Spec.Replicas, oldSet).List()
+	for _i := len(podOrdinals) - 1; _i >= 0; _i-- {
+		i := podOrdinals[_i]
+		podName := DMMasterPodName(dcName, i)
+		pod, err := mu.podLister.Pods(ns).Get(podName)
+		if err != nil {
+			return fmt.Errorf("gracefulUpgrade: failed to get pods %s for cluster %s/%s, error: %s", podName, ns, dcName, err)
+		}
+
+		revision, exist := pod.Labels[apps.ControllerRevisionHashLabelKey]
+		if !exist {
+			return controller.RequeueErrorf("tidbcluster: [%s/%s]'s dm-master pod: [%s] has no label: %s", ns, dcName, podName, apps.ControllerRevisionHashLabelKey)
+		}
+
+		if revision == dc.Status.Master.StatefulSet.UpdateRevision {
+			if member, exist := dc.Status.Master.Members[podName]; !exist || !member.Health {
+				return controller.RequeueErrorf("tidbcluster: [%s/%s]'s dm-master upgraded pod: [%s] is not ready", ns, dcName, podName)
+			}
+			continue
+		}
+
+		if controller.PodWebhookEnabled {
+			setUpgradePartition(newSet, i)
+			return nil
+		}
+
+		return mu.upgradeMasterPod(dc, i, newSet)
+	}
+
+	return nil
+}
+
+func (mu *masterUpgrader) upgradeMasterPod(dc *v1alpha1.DMCluster, ordinal int32, newSet *apps.StatefulSet) error {
+	ns := dc.GetNamespace()
+	dcName := dc.GetName()
+	upgradePodName := DMMasterPodName(dcName, ordinal)
+	if dc.Status.Master.Leader.Name == upgradePodName && dc.MasterStsActualReplicas() > 1 {
+		err := mu.evictMasterLeader(dc, upgradePodName)
+		if err != nil {
+			klog.Errorf("dm-master upgrader: failed to evict dm-master %s's leader: %v", upgradePodName, err)
+			return err
+		}
+		klog.Infof("dm-master upgrader: evict dm-master %s's leader successfully", upgradePodName)
+		return controller.RequeueErrorf("dmcluster: [%s/%s]'s dm-master member: evicting [%s]'s leader", ns, dcName, upgradePodName)
+	}
+
+	setUpgradePartition(newSet, ordinal)
+	return nil
+}
+
+func (mu *masterUpgrader) evictMasterLeader(dc *v1alpha1.DMCluster, podName string) error {
+	return controller.GetMasterPeerClient(mu.masterControl, dc, podName).EvictLeader()
```

Comment:
Why evict leader needs the peer client?

Reply:
Because dm-master doesn't support specify the leader to be evicted.